enhancements/multi-arch/multiarch-manager-operator.md

```diff
@@ -0,0 +1,703 @@
+`PodPlacementController` does not override the behavior previously pushed by
+other stakeholders, leaving it as complies with KEP-3838[^2].
+
+The multiarch-manager-operator will be available as part of the OperatorHub, as
+a secondary operator managed by the Operator Lifecycle Manager (OLM). It will be
+a cluster-scoped operator that only cluster admins can install and configure.
+
+Its development is expected to ship in the following phases:
+
+- Phase 1 - Technology Preview (TP):
+  - The pod placement controller will be supported in High Availability (HA)
+    mode
+  - The webhook will be a stateless component
+  - The pod placement controller will support namespace filtering to skip pods
+    that should not be mutated
+  - No support for customized network scenarios (restricted, proxied,
+    disconnected...)
+  - No TLS verification for the registries as we are only fetching metadata
+  - No caching facilities for the image metadata
+  - No graceful uninstallation support: any remaining gated pods have to be
+    manually patched to resume scheduling after uninstallation
+- Phase 2 - General Availability (GA):
+  - support for customized network scenarios (restricted, proxied,
+    disconnected...) 
+  - implement caching facilities for the image metadata
+- Phase 3:
+  - TLS verification for the registries
+  - Future performance optimizations
+
+### Workflow Description
+The following section will highlight the main workflows the operator will be
+responsible for from a high-level user perspective.
+
+#### Operator installation
+1. The cluster-admin submits a Subscription object to install the
+   multiarch-managed-operator, either via CLI or the OperatorHub
+2. The cluster admin submits the default `PodPlacementConfig` Custom Resource
+   (CR) adding their specific configuration related to initializing the
+   multiarch-manager-operator's components; the cluster admin can customize the
+   CR by excluding namespaces through regexes/globs, a list of configmaps for
+   registries that use self signed certs (if any) and customize logging levels
+4. The operator reconciles the CR. Its information supports the
+   `PodPlacementController` controller and the `SchedulingGateMutatingWebhook`
+   mutating admission webhook
+
+#### Pod Admission: gating
+1. The user submits a pod
+2. The mutating admission webhook is queried and decides whether to add the
+   scheduling gate to the pod, based on the namespaces exclusion list
+3. The pod is admitted and persisted
+4. The pod is put on hold by the scheduler until all the scheduling gates are
+   removed
+
+#### NodeAffinity patching of gated pods
+1. The `PodPlacementController`, watching for pod events, gets notified of a pod
+   event
+2. If the multiarch-manager-operator's gate is not present, the pod is ignored
+3. If the gate is present, a set of unique images is built from the pod spec
+4. Each image is inspected to gather the supported architectures set
```

Comment:
how does the controller choose, locate, and access the correct pull-secret to access the image?

Reply:
The controller would look at the `ImagePullSecrets` field in the pod spec, lookup the corresponding secret in the pod namespace (which it has access to through the rbac rules) and uses that to access the image. In case this is not present, it falls back to the global pull secret in the openshift-config namespace

Reply:
those aren't the only sources of pullsecrets, the secret can also be chosen simply by looking at what pullsecrets exist in the pod's NS, iirc.

Reply:
i'll have to look into this but I guess the ImagePullSecret is controlled by the SA - if there is one associated to the pod, or if there is none at all (and no ImagePullSecret explicitly specified by the pod), it should come from the default SA in the namespace.

Reply:
the behavior has been documented in the latest commit after some prototyping.